File: src/content/docs/topics/cs/coding-problems/stack/150-evaluate-reverse-polish-notation-approach2.go
package main

import (
	"fmt"
	"strconv"
)

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func evalRPN(tokens []string) int {
	stack := []int{}
	for _, tok := range tokens {
		switch tok {
		case "+", "-", "*", "/":
			b := stack[len(stack)-1]
			a := stack[len(stack)-2]
			stack = stack[:len(stack)-2]
			var res int
			switch tok {
			case "+":
				res = a + b
			case "-":
				res = a - b
			case "*":
				res = a * b
			case "/":
				res = int(float64(a) / float64(b)) // truncate toward zero
			}
			stack = append(stack, res)
		default:
			n, _ := strconv.Atoi(tok)
			stack = append(stack, n)
		}
	}
	return stack[0]
}

func runTests() {
	assert(evalRPN([]string{"2", "1", "+", "3", "*"}) == 9)
	assert(evalRPN([]string{"4", "13", "5", "/", "+"}) == 6)
	assert(evalRPN([]string{"10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"}) == 22)
	assert(evalRPN([]string{"3"}) == 3)
	assert(evalRPN([]string{"6", "2", "/"}) == 3)
	assert(evalRPN([]string{"7", "2", "/"}) == 3)
	assert(evalRPN([]string{"-7", "2", "/"}) == -3)
	fmt.Println("all tests pass")
}

func main() { runTests() }

File: src/content/docs/topics/cs/coding-problems/stack/150-evaluate-reverse-polish-notation-approach2.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}

final class Solution {
    func evalRPN(_ tokens: [String]) -> Int {
        let operators: [String: (Int, Int) -> Int] = [
            "+": { $0 + $1 }, "-": { $0 - $1 }, "*": { $0 * $1 }, "/": { $0 / $1 },
        ]
        var stack: [Int] = []
        for token in tokens {
            if let number = Int(token) { stack.append(number); continue }
            let right = stack.removeLast()
            let left = stack.removeLast()
            if let operation = operators[token] { stack.append(operation(left, right)) }
        }
        return stack.last ?? 0
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:012ae168f0fd05bf768ca351f1d8fc6657b439da9945b1499da0a6453a497107
    expectEqual(Solution().evalRPN(["2", "1", "+", "3", "*"]), 9, "multiply-sum")
    expectEqual(Solution().evalRPN(["4", "13", "5", "/", "+"]), 6, "division-and-addition")
    expectEqual(Solution().evalRPN(["7", "-3", "/"]), -2, "truncate-toward-zero")
    expectEqual(Solution().evalRPN(["42"]), 42, "single-number")
    // EXCLUDED_VECTOR empty-expression: [[]] | A valid expression contains at least one token.
    // EXCLUDED_VECTOR missing-operand: [["2","+"]] | Every operator requires two preceding operands.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
